File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Doms/Jpl/JplMessage.cs
using System.Text.Json.Serialization;

namespace FccDesktopAgent.Core.Adapter.Doms.Jpl;

/// <summary>
/// JPL protocol message with name, optional sub-code, and data map.
/// Serialized as JSON inside a binary STX/ETX frame by <see cref="JplFrameCodec"/>.
/// </summary>
public sealed record JplMessage(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("subCode")] int? SubCode = null,
    [property: JsonPropertyName("data")] IReadOnlyDictionary<string, string>? Data = null);

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/doms/jpl/JplMessage.kt
package com.fccmiddleware.edge.adapter.doms.jpl

import kotlinx.serialization.Serializable

/** JPL protocol message: every message has a name, optional sub-code, and JSON data map. */
@Serializable
data class JplMessage(
    val name: String,
    val subCode: Int? = null,
    val data: Map<String, String> = emptyMap(),
) {
    override fun toString(): String {
        val redactedData = data.mapValues { (key, value) ->
            if (isSensitiveKey(key)) {
                REDACTED
            } else {
                value
            }
        }
        return "JplMessage(name=$name, subCode=$subCode, data=$redactedData)"
    }

    companion object {
        private const val REDACTED = "[REDACTED]"
        private val sensitiveKeyMarkers = listOf("accesscode", "secret", "password")

        private fun isSensitiveKey(key: String): Boolean =
            sensitiveKeyMarkers.any { marker -> key.contains(marker, ignoreCase = true) }
    }
}
